src/Storages/Distributed/DistributedBlockOutputStream.cpp

```diff
@@ -628,6 +628,8 @@ void DistributedBlockOutputStream::writeToShard(const Block & block, const std::
     /// and keep monitor thread out from reading incomplete data
     std::string first_file_tmp_path;
 
+    storage.delayInsertOrThrowIfNeeded();
```

Comment:
Do I understand correctly that we potentially can pass some blocks from multi-block insert and throw in the middle?

Reply:
Yes, let's move it into `writePrefix` to avoid this.